### PR TITLE
Marshal v2 API's compatible with Golang Protobuf v2 (1/n)

### PR DESCRIPTION
**Pull request by gandhikrishna**
- [* ] Description and context for reviewers: one partner, one stranger

Gogo is today used in yarpc as the core library for proto code generation and for proto runtime methods. This pull request aims to make yarpc compatible with golang/protobuf v2 APIs.

This PR introduces a new Marshal v2 API's compatible with Golang Protobuf v2. These are internal APIs and constructors used with yarpc generated protobuf stubs.

### CODE REVIEW

File: encoding/protobuf/v2/marshal.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+import (
+	"bytes"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/encoding/protobuf"
+	"go.uber.org/yarpc/internal/bufferpool"
+	"go.uber.org/yarpc/yarpcerrors"
+	"io"
+	"sync"
```
Comment (gopalchouhan) [inline encoding/protobuf/v2/marshal.go:32]: Correct the import ordering.There should be two import groups:

Standard library
//new line
Everything else

File: encoding/protobuf/v2/marshal.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+)
+
+const (
+	// Encoding is the name of this encoding.
+	Encoding transport.Encoding = "proto"
```
Comment (gopalchouhan) [inline encoding/protobuf/v2/marshal.go:37]: Just to confirm, we plan to use this exported constant outside this package,right?

File: encoding/protobuf/v2/marshal.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+func unmarshalJSON(body []byte, message proto.Message, codec *codec) error {
+	return codec.jsonUnmarshaler.Unmarshal(bytes.NewReader(body), message)
+}
+
+func marshal(encoding transport.Encoding, message proto.Message, codec *codec) ([]byte, func(), error) {
+	switch encoding {
+	case protobuf.Encoding:
+		return marshalProto(message, codec)
+	case protobuf.JSONEncoding:
```
Comment (gopalchouhan) [inline encoding/protobuf/v2/marshal.go:102]: Shoudn't we be using _Encoding_ and _JSONEncoding_ constants defined in this file?

**Review by gopalchouhan — COMMENTED** [review]
Overall the change looks the duplicate copy of encoding/prototobuf/marshal.go. Given that no active development in the gogo protobuf namespace and it is possible those interface might go out of sync(or must have already gone out of sync), hence it is better to dup the code to keep it cleaner.

File: encoding/protobuf/v2/marshal.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
```
[the remark this answers is not in the document]
Comment (gandhikrishna) [reply encoding/protobuf/v2/marshal.go:1]: Once this diff is merged, post that I will update the timestamp for whole repo.

File: encoding/protobuf/v2/marshal.go (hunk 2 above)
Comment (gandhikrishna) [reply encoding/protobuf/v2/marshal.go:37]: Yes will be used in unmarshalling pkg's also with the coming diffs.

File: encoding/protobuf/v2/marshal_test.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/yarpcerrors"
+)
+
+func TestUnhandledEncoding(t *testing.T) {
+	assert.Equal(t, yarpcerrors.CodeInternal,
+		yarpcerrors.FromError(unmarshal(transport.Encoding("foo"), bytes.NewReader([]byte("foo")), nil, newCodec(nil))).Code())
+	_, _, err := marshal(transport.Encoding("foo"), nil, newCodec(nil))
+	assert.Equal(t, yarpcerrors.CodeInternal, yarpcerrors.FromError(err).Code())
+}
```
[the remark this answers is not in the document]
Comment (gandhikrishna) [reply encoding/protobuf/v2/marshal_test.go:37]: Other valid use cases will require proto messages. Generation of proto messages with v2 is not included in this diff. With subsequent diff, you will find the extended test cases.

File: encoding/protobuf/v2/marshal_test.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
```
[the remark this answers is not in the document]
Comment (gopalchouhan) [reply encoding/protobuf/v2/marshal_test.go:1]: As Krishna mentioned earlier, this will be done for entire repo together.

**Review by gopalchouhan — APPROVED** [review]

**Pull request merged**